hazelcast/src/main/java/com/hazelcast/core/Pipeline.java

```diff
@@ -0,0 +1,145 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.core;
+
+
+import com.hazelcast.spi.annotation.Beta;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Semaphore;
+
+import static com.hazelcast.util.Preconditions.checkNotNull;
+import static com.hazelcast.util.Preconditions.checkPositive;
+
+
+/**
+ * The Pipeline can be used to speed up requests. It is build on top of asynchronous
+ * requests like e.g. {@link IMap#getAsync(Object)} or any other asynchronous call.
+ *
+ * The main purpose of the pipeline is to control the number of concurrent requests
+ * when using asynchronous invocations. This can be done by setting the depth using
+ * the constructor. So you could set the depth to e.g 100 and do 1000 calls. That means
+ * that at any given moment, there will only be 100 concurrent requests.
+ *
+ * It depends on the situation what the optimal depth (number of invocations in
+ * flight) should be. If it is too high, you can run into memory related problems.
+ * If it is too low, it will provide little or no performance advantage at all. In
+ * most cases a pipeline a a few hundred map/cache puts/gets should not lead to any
+ * problems. For testing purposes we frequently have a pipeline of 1000 or more
+ * concurrent requests.
+ *
+ * The Pipeline can't be used for transaction purposes. So you can't create a
+ * Pipeline, add a set of asynchronous request and then not call {@link #results()}
+ * to prevent executing these requests. Invocations can be executed before the
+ * {@link #results()} is called.
+ *
+ * Pipelines can be used by both clients or members.
+ *
+ * The Pipeline isn't threadsafe. So only a single thread should add requests to
+ * the pipeline and wait for results.
+ *
+ * Currently all {@link ICompletableFuture} and their responses are stored in the
+ * Pipeline. So be careful executing a huge number of request with a single pipeline
+ * because it can lead to a huge memory bubble. In this cases it is better to
+ * periodically, after waiting for completion, to replace the pipeline by a new one.
+ * In the future we might provide this as an out of the box experience, but currently
+ * we do not.
+ *
+ * The Pipeline has been marked as Beta since we need to see how the API needs to
+ * evolve. But there is no problem using it in production. We use similar techniques
+ * to achieve high performance.
+ *
+ * @param <E>
+ */
+@Beta
+public class Pipeline<E> {
```

Comment:
Are there any other planned additions to Pipeline API? If no, why are we adding this to core API? This is trivial to implement for users themselves.

Reply:
We need to see how this API should evolve.

I think it should be added to core because it can make a very big impact on performance. It should be an out of the box and easy to find functionality.